allow use of secure gRPC connections between agent and collector

ghouscht opened the pull request:
# Which problem is this PR solving?

Resolves #1310

## Short description of the changes

Implementation for a secure gRPC connection between the agent and the collector.

ghouscht commented:
Looks like an unrelated test is now failing in CI: `TestSaramaConsumerWrapper_start_Messages`. Can't reproduce it on my machine, is this test known to be flaky/broken?

yurishkuro commented:
possibly, I restarted the build to see.

yurishkuro commented:
Confirmed flaky test - #1314

yurishkuro commented:
@ghouscht what is your action plan with this PR? Are you going to add the equivalent settings to the collector?

ghouscht commented:
Yes, today I wrote some very basic unit tests for the agent part (just flags and ProxyBuilder). 
Will have a look at the collector code tomorrow to add those settings there to. The only thing left then, is some sort of e2e test, but don't know at the moment how to do that...

ghouscht commented:
Code for the collector part now also finished. Tested it on my local machine and seems to work well. 

things left TODO:
- agent: rewrite unit test `TestProxyBuilder`with custom CA (open for ideas) possible idea: write a cert to a tempfile and use this file to test the `NewCollectorProxy` method with a real CA or is the test like it is now ok?
- collector: do we need a custom error message if the user starts the collector with `--collector.grpc.tls` but does not provide a cert and key? Or just stay with the current error which is like: `"Could not start gRPC collector","error":"failed to load TLS keys: open : no such file or directory"` (not ideal in my opinion)?
- E2E test? how?

yurishkuro commented:
loading certs from a file in /tmp might be a good way to unit test the initialization paths

ghouscht commented:
Added [pkg/testutils/certificate.go](https://github.com/jaegertracing/jaeger/pull/1311/files#diff-55ee2fc4b2ba1290248a7b20168aec0d), maybe a bit overkill at the moment. But this method could be used too for an e2e/integration test where a cert and key is needed.

Do you think there is anything left which needs to be done? E2E test or tests for the initialization paths of the collector in [cmd/collector/main.go](https://github.com/jaegertracing/jaeger/blob/master/cmd/collector/main.go) (currently no tests at all for this part of the code)?

yurishkuro commented on pkg/testutils/certificate.go:
```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2019 The Jaeger Authors.
```
It doesn't feel like we should have this code (which is also hard to test). Go already comes with `crypto/tls/generate_cert.go` that we can run (there's also installable version https://github.com/odeke-em/ssg).

ghouscht replied:
sure, removed that part of the PR

yurishkuro commented:
we do have integration tests running binaries, `crossdock/docker-compose.yml`, but it's hard to extend. In fact, we still haven't switched it to use gRPC.

ghouscht commented:
Cleaned up the PR and removed the certificate stuff. `TestSaramaConsumerWrapper_start_Messages` is failing again... (this time with a different error I think so).

benley commented:
You'll probably want to add a ssl trusted roots bundle to the jaeger-agent docker image, similar to what already exists in [jaeger-collector's Dockerfile](https://github.com/jaegertracing/jaeger/blob/c4bb6b524f6be69b3268757f0c2b7c43e4ec7b4b/cmd/collector/Dockerfile#L1-L6):

```diff
--- a/cmd/agent/Dockerfile
+++ b/cmd/agent/Dockerfile
@@ -1,5 +1,10 @@
+FROM alpine:latest as certs
+RUN apk add --update --no-cache ca-certificates
+
 FROM scratch
 
+COPY --from=certs /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
+
 EXPOSE 5775/udp 6831/udp 6832/udp 5778
 COPY agent-linux /go/bin/
-ENTRYPOINT ["/go/bin/agent-linux"]
+ENTRYPOINT ["/go/bin/agent-linux", "--reporter.grpc.tls.ca=/etc/ssl/certs/ca-certificates.crt"]
```

ghouscht commented:
@yurishkuro how to proceed now with this PR?

pavolloffay commented on cmd/agent/app/reporter/grpc/flags.go:
```diff
@@ -22,19 +22,28 @@ import (
 // AddFlags adds flags for Options.
 func AddFlags(flags *flag.FlagSet) {
 	flags.String(collectorHostPort, "", "(experimental) Comma-separated string representing host:port of a static list of collectors to connect to directly.")
+	flags.Bool(collectorTLS, false, "(experimental) Enable TLS.")
```
We have removed experimental flags from grpc

ghouscht replied:
Ok, will rebase my PR to master and change this in the code..

ghouscht commented:
Looks like an unrelated test is failing again? What about coverage, is it ok like this?

pavolloffay commented:
The coverage should remain at 100%.

ghouscht commented:
Ok, catching the first miss [here](https://codecov.io/gh/jaegertracing/jaeger/pull/1311/diff?src=pr&el=tree#D5-53) could be hard, or at least I don't know how to do this, any ideas?

The second miss [here](https://codecov.io/gh/jaegertracing/jaeger/pull/1311/diff?src=pr&el=tree#D5-60) shouldn't be a problem, will do this in a few minutes.

yurishkuro commented:
@ghouscht I fixed coverage by adding one more test (#1391). I can get it approved and merged tomorrow, or you can merge it into your branch, and then we can merge this PR.

NB: in the future I recommend not committing directly to master in your fork, but creating a branch (as I did in #1391). Had you done that, I would've been able to push these extra commits directly into your branch and update this PR automatically.

ghouscht commented:
I'm out  of the office today. Will merge the changes into my branch tomorrow and squash the commits together.

yurishkuro commented:
merged as #1391
